### PR TITLE
fix complex types returning UNKNOWN as their SQL type inference

### PR DESCRIPTION
**Pull request by zachjsh**
### Description

Fixes issue where several aggregators are returning UNKNOWN or OTHER as their SQL type inference. Fixing this will allow these types do be validated against when defined as column on tables stored in the catalog.

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: sql/src/main/java/org/apache/druid/sql/calcite/expression/OperatorConversions.java [hunk 1]
```diff
@@ -779,4 +781,13 @@ public static DirectOperatorConversion druidUnaryDoubleFn(String sqlOperator, St
         druidFunctionName
     );
   }
+
+  public static SqlReturnTypeInference complexReturnTypeWithNullability(ColumnType columnType, boolean nullable)
```
[the remark this answers is not in the document]
Comment (zachjsh) [reply sql/src/main/java/org/apache/druid/sql/calcite/expression/OperatorConversions.java]: Thanks! Moved.

File: extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/theta/sql/ThetaSketchEstimateWithErrorBoundsOperatorConversion.java [hunk 2]
```diff
@@ -46,7 +46,8 @@ public class ThetaSketchEstimateWithErrorBoundsOperatorConversion implements Sql
   private static final SqlFunction SQL_FUNCTION = OperatorConversions
       .operatorBuilder(StringUtils.toUpperCase(FUNCTION_NAME))
       .operandTypes(SqlTypeFamily.ANY, SqlTypeFamily.INTEGER)
-      .returnTypeNonNull(SqlTypeName.OTHER)
+      .returnTypeInference(
+          OperatorConversions.complexReturnTypeWithNullability(SketchModule.BUILD_TYPE, false))
```
[the remark this answers is not in the document]
Comment (zachjsh) [reply extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/theta/sql/ThetaSketchEstimateWithErrorBoundsOperatorConversion.java]: Thanks fixed, and think I got all the formatting ones, let me know if I missed any.

File: extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/quantiles/sql/DoublesSketchObjectSqlAggregator.java [hunk 3]
```diff
@@ -52,7 +52,8 @@ public class DoublesSketchObjectSqlAggregator implements SqlAggregator
       OperatorConversions.aggregatorBuilder(NAME)
                          .operandNames("column", "k")
                          .operandTypes(SqlTypeFamily.ANY, SqlTypeFamily.EXACT_NUMERIC)
-                         .returnTypeNonNull(SqlTypeName.OTHER)
+                         .returnTypeInference(
+                             OperatorConversions.complexReturnTypeWithNullability(HllSketchModule.TYPE, false))
```
[the remark this answers is not in the document]
Comment (zachjsh) [reply extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/quantiles/sql/DoublesSketchObjectSqlAggregator.java]: thanks, fixed

File: extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/sql/HllSketchEstimateWithErrorBoundsOperatorConversion.java [hunk 4]
```diff
@@ -47,7 +47,8 @@ public class HllSketchEstimateWithErrorBoundsOperatorConversion implements SqlOp
       .operatorBuilder(StringUtils.toUpperCase(FUNCTION_NAME))
       .operandTypes(SqlTypeFamily.ANY, SqlTypeFamily.INTEGER)
       .requiredOperandCount(1)
-      .returnTypeNonNull(SqlTypeName.OTHER)
+      .returnTypeInference(
```
[the remark this answers is not in the document]
Comment (zachjsh) [reply extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/sql/HllSketchEstimateWithErrorBoundsOperatorConversion.java]: Thanks, fixed

**Review by jon-wei — APPROVED** [review]

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/Calcites.java [hunk 5]
```diff
@@ -100,6 +100,15 @@ public class Calcites
   public static final SqlReturnTypeInference
       ARG1_NULLABLE_ARRAY_RETURN_TYPE_INFERENCE = new Arg1NullableArrayTypeInference();
 
+  public static SqlReturnTypeInference complexReturnTypeWithNullability(ColumnType columnType, boolean nullable)
```
Comment (clintropolis) [inline sql/src/main/java/org/apache/druid/sql/calcite/planner/Calcites.java:103]: Oh, I guess we could have defined this method on `OperatorBuilder`, similar to the others https://github.com/apache/druid/blob/master/sql/src/main/java/org/apache/druid/sql/calcite/expression/OperatorConversions.java#L366

**Review by clintropolis — APPROVED** [review]

**Pull request merged**